### PR TITLE
work-item strided transforms

### PR DESCRIPTION
**Pull request by FMarno**
Adds work-item level strided ffts.
This includes support for arbitrary strides and distances for 1D transforms, layout validation, and validation tests.

## Checklist

Tick if relevant:

* [x] New files have a copyright
* [x] New headers have an include guards
* [x] API is documented with Doxygen
* [x] New functionalities are tested
* [x] Tests pass locally
* [x] Files are clang-formatted

### CODE REVIEW

File: test/unit_test/fft_test_utils.hpp [hunk 1]
```diff
@@ -95,44 +123,57 @@ struct test_params {
...
+
+    if (params.strides) {
+      ss << "__FwdStrides";
+      for (std::size_t s : params.strides.value().forward_strides) {
+        ss << "_" << s;
+      }
+      ss << "__FwdDistance_" << params.strides.value().forward_distance;
+      ss << "__BwdStrides";
+      for (std::size_t s : params.strides.value().backward_strides) {
+        ss << "_" << s;
```
Comment (AD2605) [inline test/unit_test/fft_test_utils.hpp:156]: We would need to change the test names where ever necessary for now (or probably in every test right away), and omit the LayoutIn and LayoutOut parameters.
For example in the arbitrary strided test suite, mentioning the layouts would not make sense

File: src/portfft/descriptor_validate.hpp [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    throw invalid_configuration("Invalid lengths, must have at least 1 dimension");
+  }
+  for (std::size_t i = 0; i < lengths.size(); ++i) {
+    if (lengths[i] == 0) {
+      throw invalid_configuration("Invalid lengths[", i, "]=", lengths[i], ", must be positive");
+    }
+  }
+}
+
+inline void validate_layout(const std::vector<std::size_t>& lengths, detail::layout forward_layout,
```
Comment (AD2605) [inline src/portfft/descriptor_validate.hpp:43]: I suppose the layout is also the part of the descriptor, and not the committed descriptor correct ?
In that case let's keep these checks (this and the validate_lengths) in the constructor of descriptor itself. This way we needn't build the kernels before throwing `invalid_configuration`

Comment (AD2605) [reply src/portfft/descriptor_validate.hpp:43]: After going through the remaining checks, I suppose quite a few can be moved to the descriptor. 
I am of the opinion that we should not allow one to commit an invalid configuration

File: src/portfft/dispatcher/workitem_dispatcher.hpp [hunk 3]
```diff
@@ -140,24 +144,67 @@ PORTFFT_INLINE void workitem_impl(const T* input, T* output, const T* input_imag
...
+    IdxGlobal leader_i = i - static_cast<IdxGlobal>(subgroup_local_id);
+
+    Idx n_working = sycl::min(SubgroupSize, static_cast<Idx>(n_transforms - leader_i));
+    IdxGlobal global_offset = static_cast<IdxGlobal>(n_io_reals) * leader_i;
+    IdxGlobal global_input_offset = static_cast<IdxGlobal>(input_distance_in_reals) * leader_i;
+    IdxGlobal global_output_offset = static_cast<IdxGlobal>(output_distance_in_reals) * leader_i;
+
+    // This is checking for LayoutIn is PACKED or UNPACKED but we don't actually ever launch kernels with LayoutIn
+    // as UNPACKED.
+    if (LayoutIn == detail::layout::PACKED || LayoutIn == detail::layout::UNPACKED) {
```
Comment (AD2605) [inline src/portfft/dispatcher/workitem_dispatcher.hpp:161]: Our kernels will soon become very hard to follow, with the upcoming subgroup and workgroup strided PRs.
Would it be, or rather would it be a good idea to isolate, batch_interleaved / packed / unpacked logic into different functions, and place them in a different file ?

File: src/portfft/descriptor.hpp [hunk 4]
```diff
@@ -542,67 +485,110 @@ class committed_descriptor {
...
       } else {
+        length = static_cast<Idx>(params.lengths[dimension_num]);
+
+        forward_stride = static_cast<Idx>(params.forward_strides[dimension_num]);
+        backward_stride = static_cast<Idx>(params.backward_strides[dimension_num]);
+        if (is_multi_dim) {
+          if (dimension_num == (params.lengths.size() - 1)) {
+            forward_distance = length;
+            backward_distance = length;
+            apply_scale = detail::apply_scale_factor::APPLIED;
```
Comment (AD2605) [inline src/portfft/descriptor.hpp]: I would have a small preference towards the way the it was being decided upon whether or not to skip scaling in the case of multi-dimensional transforms, as the `is_multi_dim` check is not required. It would also reduce the number of cases we need to handle inside `set_spec_const_driver`

File: src/portfft/descriptor_validate.hpp [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+inline void validate_lengths(const std::vector<std::size_t>& lengths) {
+  if (lengths.empty()) {
+    throw invalid_configuration("Invalid lengths, must have at least 1 dimension");
+  }
+  for (std::size_t i = 0; i < lengths.size(); ++i) {
+    if (lengths[i] == 0) {
+      throw invalid_configuration("Invalid lengths[", i, "]=", lengths[i], ", must be positive");
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/descriptor_validate.hpp:46]: This function is called from the constructor of `committed_descriptor`, before `build_w_spec_const`. I put all the validation stuff in a separate file because it can easily be compartmentalised in its own little place, and not distract from other bits.

File: src/portfft/dispatcher/workitem_dispatcher.hpp [hunk 6]
```diff
@@ -255,18 +305,43 @@ PORTFFT_INLINE void workitem_impl(const T* input, T* output, const T* input_imag
         }
       }
     }
-    if (LayoutOut == detail::layout::PACKED) {
+    // This is checking for LayoutOut is PACKED or UNPACKED but we don't actually ever launch kernels with LayoutOut
+    // as UNPACKED, but just use PACKED and check the stride.
+    if (LayoutOut == detail::layout::PACKED || LayoutOut == detail::layout::UNPACKED) {
```
[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/dispatcher/workitem_dispatcher.hpp:311]: I believe we agreed we would leave this slight oddness for now and sort it when we remove all the layout templates

File: src/portfft/dispatcher/workitem_dispatcher.hpp [hunk 7]
```diff
@@ -255,18 +305,43 @@ PORTFFT_INLINE void workitem_impl(const T* input, T* output, const T* input_imag
...
       sycl::group_barrier(global_data.sg);
       global_data.log_dump_local("computed data local memory:", loc, n_reals * n_working);
       global_data.log_message_global(__func__, "storing data from local to global memory");
-      if (storage == complex_storage::INTERLEAVED_COMPLEX) {
-        local2global<level::SUBGROUP, SubgroupSize>(global_data, loc_view, output, n_reals * n_working, local_offset,
-                                                    global_offset);
+      if (output_stride == 1 && output_distance == fft_size) {
```
[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/dispatcher/workitem_dispatcher.hpp:315]: That is because we never actually use instantiate LayoutOut == UNPACKED kernels, so its kind of check the it is actually packed. Its described in the comment directly above.
I can see why its confusing but a proper fix would be more disruptive.

File: test/unit_test/fft_test_utils.hpp (hunk 1 above)
Comment (FMarno) [reply test/unit_test/fft_test_utils.hpp:156]: Do you think this needs to be solved now, or when removing the Layout parameters?

File: src/portfft/descriptor_validate.hpp (hunk 2 above)
Comment (FMarno) [reply src/portfft/descriptor_validate.hpp]: As above, this is part of the constructor for `committed_descriptor`, so all the code in descriptor_validate.hpp is called before we try to build any kernel bundles.
The committed_descriptor class is big enough in my opinion!

File: src/portfft/descriptor.hpp [hunk 8]
```diff
@@ -32,6 +32,7 @@
 #include "common/exceptions.hpp"
 #include "common/subgroup.hpp"
 #include "defines.hpp"
+#include "descriptor_validate.hpp"
```
[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/descriptor.hpp:35]: I'm not so bothered either way, but why do you think that?

File: src/portfft/descriptor_validate.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (AD2605) [reply src/portfft/descriptor_validate.hpp:46]: Its perfectly alright where the functions are defined. Its about where they are called from.  Combining this with [this comment](https://github.com/codeplaysoftware/portFFT/pull/136#discussion_r1472731845), what I am trying to discuss, is that, since a user would specify the layout configuration in the descriptor itself, should we let someone commit the descriptor at all, if the provided layout configuration is invalid

File: src/portfft/specialization_constant.hpp [hunk 9]
```diff
@@ -26,12 +26,16 @@
...
 
-namespace portfft {
-namespace detail {
+namespace portfft::detail {
 
 constexpr static sycl::specialization_id<Idx> SpecConstFftSize{};
 constexpr static sycl::specialization_id<Idx> SpecConstNumRealsPerFFT{};
 constexpr static sycl::specialization_id<Idx> SpecConstWIScratchSize{};
+constexpr static sycl::specialization_id<Idx> SpecConstInputStride{};
+constexpr static sycl::specialization_id<Idx> SpecConstOutputStride{};
```
[the remark this answers is not in the document]
Comment (AD2605) [reply src/portfft/specialization_constant.hpp:36]: Thanks

File: test/unit_test/fft_test_utils.hpp (hunk 1 above)
Comment (AD2605) [reply test/unit_test/fft_test_utils.hpp:156]: I would suggest at least in the strided tests, where layout parameters would not make any sense at all, should be omitted in the test name (hence  the "where ever necessary for now"). Further down the line, every test name would specify the layout in terms of stride and distance

File: src/portfft/descriptor_validate.hpp (hunk 2 above)
Comment (AD2605) [reply src/portfft/descriptor_validate.hpp]: Combining this with [this])(https://github.com/codeplaysoftware/portFFT/pull/136#discussion_r1473326926)

File: src/portfft/descriptor_validate.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (AD2605) [reply src/portfft/descriptor_validate.hpp:46]: Another thing to consider over here might be how it would interact with our oneMKL integration

File: src/portfft/dispatcher/workitem_dispatcher.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (AD2605) [reply src/portfft/dispatcher/workitem_dispatcher.hpp:311]: No, I suppose we had discussed that we will not be using `Layout` template parameters in the kernels strided support is being extended to. The template parameters would just exist till we extend support to all levels, before being removed.

Only stride and distance is to be used inside the kernel to determine the layout

File: src/portfft/descriptor_validate.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/descriptor_validate.hpp:46]: So are you suggesting that the call to `detail::validate::validate_descriptor` should be [here](https://github.com/codeplaysoftware/portFFT/blob/027cc4189beedb2fa756bd1ecd1b8ff6cea189e6/src/portfft/descriptor.hpp#L1606) instead?

[the remark this answers is not in the document]
Comment (AD2605) [reply src/portfft/descriptor_validate.hpp:46]: yes, I do not suppose we should reach the constructor of committed_descriptor if the layout is illegal

File: src/portfft/common/transfers.hpp [hunk 10]
```diff
@@ -85,7 +85,7 @@ PORTFFT_INLINE void copy_group(detail::global_data_struct<1> global_data, ViewSr
   static_assert(Level == detail::level::SUBGROUP || Level == detail::level::WORKGROUP,
                 "Only subgroup and workgroup level supported");
   static_assert(!detail::is_view_multidimensional<ViewSrc>() && !detail::is_view_multidimensional<ViewDst>(),
-                "This overload of copy_wi expects one-dimensional view arguments!");
+                "This overload of copy_group expects one-dimensional view arguments!");
```
[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/common/transfers.hpp]: Done in https://github.com/codeplaysoftware/portFFT/pull/136/commits/545d2dc0be8f19148a8ccbec84283c0e9cf592cc

**Review by t4c1 — CHANGES_REQUESTED** [review]

File: src/portfft/common/logging.hpp [hunk 11]
```diff
@@ -151,8 +141,9 @@ struct global_data_struct {
    */
   template <typename TFirst, typename... Ts>
   __attribute__((always_inline)) inline void log_message_impl(TFirst message, Ts... other_messages) {
-    s << message << " ";
-    log_message_impl(other_messages...);
+    s << message;
+    ((s << ' ' << other_messages), ...);
+    s << '\n' << sycl::stream_manipulator::flush;
```
[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/common/logging.hpp]: I've removed this code this it wasn't really necessary for the PR, but also this would add a space in front of all messages, which disproportionately upsets me!

File: test/common/reference_data_wrangler.hpp [hunk 12]
```diff
@@ -69,13 +103,13 @@ std::vector<T> transpose(const std::vector<T>& in, std::size_t dft_len, std::siz
...
-template <portfft::direction Dir, portfft::complex_storage Storage, typename Scalar, portfft::domain Domain,
-          typename PaddingT>
+template <portfft::direction Dir, portfft::complex_storage Storage, typename Scalar, portfft::domain Domain>
 auto gen_fourier_data(portfft::descriptor<Scalar, Domain>& desc, portfft::detail::layout layout_in,
-                      portfft::detail::layout layout_out, PaddingT padding_value) {
+                      portfft::detail::layout layout_out, float padding_value) {
   constexpr bool IsRealDomain = Domain == portfft::domain::REAL;
   constexpr bool IsForward = Dir == portfft::direction::FORWARD;
   constexpr bool IsInterleaved = Storage == portfft::complex_storage::INTERLEAVED_COMPLEX;
+  constexpr bool debug_input = false;
```
[the remark this answers is not in the document]
Comment (FMarno) [reply test/common/reference_data_wrangler.hpp:112]: Personally, I've found it quite helpful. I've become quite familiar with the output from ascending numbers!
Also if you want to override the debug input then python code is closer to what you might want e.g replace the arange with a repeated value.

File: src/portfft/descriptor.hpp [hunk 13]
```diff
@@ -565,69 +510,86 @@ class committed_descriptor {
...
+      const auto conjugate_on_store =
+          is_backward && is_final_factor ? detail::complex_conjugate::APPLIED : detail::complex_conjugate::NOT_APPLIED;
+      const auto apply_scale = is_final_factor && is_final_dim ? detail::apply_scale_factor::APPLIED
+                                                               : detail::apply_scale_factor::NOT_APPLIED;
+
+      Idx length{};
+      Idx forward_stride{};
+      Idx backward_stride{};
+      Idx forward_distance{};
+      Idx backward_distance{};
```
[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/descriptor.hpp]: addressed here: https://github.com/codeplaysoftware/portFFT/pull/136/commits/7271c29ae1b976b8d86de3ce12003833e2bb246b

File: src/portfft/descriptor_validate.hpp [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+#include "common/exceptions.hpp"
+#include "enums.hpp"
+#include "utils.hpp"
+
+namespace portfft::detail::validate {
+
+namespace detail {
+
+/**
+ * Throw an exception if the lengths are invalid when looked at in isolation.
```
[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/descriptor_validate.hpp:35]: addressed in https://github.com/codeplaysoftware/portFFT/pull/136/commits/3c58281a4037592ec9fdf0e68fc6163ab7e0b1f4

File: src/portfft/dispatcher/workitem_dispatcher.hpp [hunk 15]
```diff
@@ -256,18 +306,43 @@ PORTFFT_INLINE void workitem_impl(const T* input, T* output, const T* input_imag
...
```
[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/dispatcher/workitem_dispatcher.hpp:314]: addressed here: https://github.com/codeplaysoftware/portFFT/pull/136/commits/8c8393de300609ed88c084375a70085d9d395270

File: src/portfft/descriptor.hpp [hunk 16]
```diff
@@ -542,67 +485,110 @@ class committed_descriptor {
...
     Scalar scale_factor = compute_direction == direction::FORWARD ? params.forward_scale : params.backward_scale;
-    detail::apply_scale_factor scale_factor_applied = detail::apply_scale_factor::APPLIED;
-    bool is_compatible = true;
-    if (skip_scaling) {
-      scale_factor_applied = detail::apply_scale_factor::NOT_APPLIED;
-    }
     std::size_t counter = 0;
-    auto conjugate_on_load = detail::complex_conjugate::NOT_APPLIED;
-    auto conjugate_on_store = detail::complex_conjugate::NOT_APPLIED;
+    Idx remaining_factors_prod = static_cast<Idx>(params.get_flattened_length());
```
[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/descriptor.hpp]: Sure, done here https://github.com/codeplaysoftware/portFFT/pull/136/commits/7271c29ae1b976b8d86de3ce12003833e2bb246b

File: src/portfft/descriptor.hpp [hunk 17]
```diff
@@ -453,23 +393,26 @@ class committed_descriptor {
    * @param num_factors total number of factors of the committed size, set as a spec constant
    */
   void set_spec_constants(detail::level top_level, sycl::kernel_bundle<sycl::bundle_state::input>& in_bundle,
-                          std::size_t length, const std::vector<Idx>& factors,
-                          detail::elementwise_multiply multiply_on_load, detail::elementwise_multiply multiply_on_store,
+                          Idx length, const std::vector<Idx>& factors, detail::elementwise_multiply multiply_on_load,
```
[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/descriptor.hpp:404]: I think the change goes along nicely with the changes to set_spec_constant_driver. I have reverted some of the less obviously connected changes though.

File: src/portfft/descriptor_validate.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/descriptor_validate.hpp:48]: Sounds pretty reasonable to me. https://github.com/codeplaysoftware/portFFT/pull/136/commits/f54b9b773e5e54de69e3eaffc109f68b42a8be33

File: src/portfft/dispatcher/workitem_dispatcher.hpp [hunk 18]
```diff
@@ -141,24 +145,67 @@ PORTFFT_INLINE void workitem_impl(const T* input, T* output, const T* input_imag
...
+    IdxGlobal global_input_offset = static_cast<IdxGlobal>(input_distance_in_reals) * leader_i;
+    IdxGlobal global_output_offset = static_cast<IdxGlobal>(output_distance_in_reals) * leader_i;
+
+    // This is checking for LayoutIn is PACKED or UNPACKED but we don't actually ever launch kernels with LayoutIn
+    // as UNPACKED.
+    if (LayoutIn == detail::layout::PACKED || LayoutIn == detail::layout::UNPACKED) {
+      // copy into local memory cooperatively as a subgroup, allowing coalesced memory access for when elements of a
+      // single FFT are sequential. BATCH_INTERLEAVED skips this step and loads straight from global to registers since
+      // the sequential work-items already access sequential elements.
+      if (input_stride == 1 && input_distance == fft_size) {
```
Comment (Rbiessy) [inline src/portfft/dispatcher/workitem_dispatcher.hpp:166]: I still don't like that we nest different `if` conditions that test the layout. The first `if (LayoutIn == detail::layout::PACKED || LayoutIn == detail::layout::UNPACKED)` would make sense if the 2 layouts had some common logic but there is almost nothing currently. It would make more sense to reduce the nested-ness like:
```
if (layout_in == PACKED) {
...
} else if (layout_in == BATCH_INTERLEAVED) {
...
} else {
...
}
```
That would only work if you already make the change to define `layout_in` inside the kernel.

File: src/portfft/enums.hpp [hunk 19]
```diff
@@ -43,6 +43,20 @@ enum class pad { DONT_PAD, DO_PAD };
 
 enum class level { WORKITEM, SUBGROUP, WORKGROUP, GLOBAL };
 
+constexpr const char* level_to_string(level l) {
```
Comment (Rbiessy) [inline src/portfft/enums.hpp:46]: Are you using this somewhere?

File: test/unit_test/fft_test_utils.hpp [hunk 20]
```diff
@@ -72,6 +98,7 @@ struct test_params {
   std::optional<double> backward_scale;
   std::optional<std::size_t> forward_offset;
   std::optional<std::size_t> backward_offset;
+  std::optional<layout_params> strides;
```
Comment (Rbiessy) [inline test/unit_test/fft_test_utils.hpp:101]: Could we rename this to `strides_and_distances` or `layout_params`?

File: src/portfft/dispatcher/workitem_dispatcher.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (Rbiessy) [reply src/portfft/dispatcher/workitem_dispatcher.hpp:314]: I think this relates to my other [comment](https://github.com/codeplaysoftware/portFFT/pull/136#discussion_r1478595269). I agree that this is a bit too confusing even if we expect this to be a temporary solution.

File: src/portfft/dispatcher/workitem_dispatcher.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/dispatcher/workitem_dispatcher.hpp:311]: Ok, it seems I misunderstood. I can do that.

File: src/portfft/descriptor_validate.hpp [hunk 21]
```diff
@@ -0,0 +1,271 @@
...
+ */
+inline void strides_distance_check(const std::vector<std::size_t>& lengths, std::size_t number_of_transforms,
+                                   const std::vector<std::size_t>& strides, std::size_t distance,
+                                   const std::string_view domain_str) {
+  validate_strides_distance_basic(lengths, number_of_transforms, strides, distance, domain_str);
+  if (lengths.size() > 1) {
+    strides_distance_multidim_check(lengths, number_of_transforms, strides, distance, domain_str);
+  } else {
+    strides_distance_1d_check(lengths, number_of_transforms, strides, distance, domain_str);
+  }
```
[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/descriptor_validate.hpp:207]: Thank you :D
I believe the functionality of `strides_distance_multidim_check` is cover by these lines https://github.com/codeplaysoftware/portFFT/blob/471f48c3c30b1d5c643c41005dda70c01c5266d4/src/portfft/descriptor_validate.hpp#L164-L167.
I am confident `strides_distance_1d_check` is fast and covers all the 1D cases, but generalising that to multi-dimension was not quickly obvious to me, so just kept `strides_distance_multidim_check` to cover multi-dim cases.

File: src/portfft/dispatcher/workitem_dispatcher.hpp (hunk 18 above)
Comment (FMarno) [reply src/portfft/dispatcher/workitem_dispatcher.hpp:166]: I was just not keen because it would take more time, but if we have time then I'll happily do it.

File: test/unit_test/instantiate_fft_tests.hpp [hunk 22]
```diff
@@ -211,18 +220,108 @@ INSTANTIATE_TEST_SUITE_P(OffsetsMDErrorRegressionTest, FFTTest,
...
+                         test_params_print());
+
+INSTANTIATE_TEST_SUITE_P(workItemStridedOOPInOrder, FFTTest,
+                         ::testing::ConvertGenerator<layout_param_tuple>(::testing::Combine(
+                             oop_unpacked_unpacked_layout, both_directions, complex_storages,
+                             ::testing::Values(1, 3, 33000ul),
+                             ::testing::Values(layout_params{{3}, {4}, {7}}, layout_params{{8}, {11}, {2}},
+                                               layout_params{{9}, {3}, {4}, 30, 40}))),
+                         test_params_print());
+INSTANTIATE_TEST_SUITE_P(workItemStridedOOPLikeBatchInterleaved, FFTTest,
```
[the remark this answers is not in the document]
Comment (FMarno) [reply test/unit_test/instantiate_fft_tests.hpp:244]: I thought I had left a comment but I guess I must have accidentally deleted it.
The LikeBatchInterleaved configurations will write to the same memory location twice if the number of batches is greater than the stride.

[the remark this answers is not in the document]
Comment (FMarno) [reply test/unit_test/instantiate_fft_tests.hpp]: comment clarified in https://github.com/codeplaysoftware/portFFT/pull/136/commits/cdf802c3685a04dc5d9ed6a728ba873db0724280

File: src/portfft/enums.hpp (hunk 19 above)
Comment (FMarno) [reply src/portfft/enums.hpp]: no, woops https://github.com/codeplaysoftware/portFFT/pull/136/commits/cdf802c3685a04dc5d9ed6a728ba873db0724280

File: test/unit_test/instantiate_fft_tests.hpp [hunk 23]
```diff
@@ -211,18 +220,107 @@ INSTANTIATE_TEST_SUITE_P(OffsetsMDErrorRegressionTest, FFTTest,
...
+    test_params_print());
+
+// clang-format off
+// Arbitrary interleaved FFT test suites
+// The strides and distances are set so that no elements overlap but there are no single continuous dimension in memory either.
+// This configuration is impractical but technically valid. For instance for n_batches=4, fft_size=4, stride=4, distance=3:
+// Index in memory:     0    1    2    3    4    5    6    7    8    9    10   11   12   13   14   15   16   17   18   19   20   21
+// Batch and FFT index: b0i0           b1i0 b0i1      b2i0 b1i1 b0i2 b3i0 b2i1 b1i2 b0i3 b3i1 b2i2 b1i3      b3i2 b2i3           b3i3
+// clang-format on
+INSTANTIATE_TEST_SUITE_P(workItemStridedArbitraryInterleaved, FFTTest,
```
Comment (Rbiessy) [inline test/unit_test/instantiate_fft_tests.hpp:272]: Do you have tests to make sure that we allow configurations that read from the same index? Similar to `OverlapRead` here https://github.com/codeplaysoftware/portFFT/blob/romain/stride_distance/test/unit_test/instantiate_fft_tests.hpp#L224

File: test/unit_test/instantiate_fft_tests.hpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (Rbiessy) [reply test/unit_test/instantiate_fft_tests.hpp]: Hmm ok, so this is to test the descriptor validation? Small technicality but when you say `stride >= number of transforms` it is a bit ambiguous whether you mean for both inputs and outputs or just one of them, I assume you mean "at least one of them has stride >= number of transforms"?

[the remark this answers is not in the document]
Comment (FMarno) [reply test/unit_test/instantiate_fft_tests.hpp]: The idea is that I still want to test a large number of batches, but if I wanted to do that for layouts like BatchInterleaved then I would also need all of the transforms to have really large strides.
The stride only needs to be larger than the number of transforms when the distance < stride for that domain.

File: src/portfft/descriptor_validate.hpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (Rbiessy) [reply src/portfft/descriptor_validate.hpp:207]: I see what you mean. The check in `strides_distance_multidim_check` is a little smarter than this `if` check. It sorts the indices by order of `generic_strides` so it can also quickly check that batch_interleaved configurations are valid. I think the equivalent would be:
```
const std::size_t first_batch_limit = stride * fft_size;
const std::size_t first_length_limit = distance * number_of_transforms;
if ((stride <= distance && first_batch_limit <= distance) ||
    (distance <= stride && first_length_limit <= stride)) {
    return;
}
```

[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/descriptor_validate.hpp:207]: Are you sure that that is always correct?
I've added a bit to the loop so large batch interleaved layouts don't take long to check
https://github.com/codeplaysoftware/portFFT/blob/dee36f59096caa0830a422091fb5c09d8f8e7639/src/portfft/descriptor_validate.hpp#L181-L186

File: src/portfft/dispatcher/workitem_dispatcher.hpp (hunk 18 above)
Comment (FMarno) [reply src/portfft/dispatcher/workitem_dispatcher.hpp]: Sure, addressed in this commit https://github.com/codeplaysoftware/portFFT/pull/136/commits/f8a8661520105d38262d149a3686e87054807ba7, but its probably easier to look at the diff from main

File: src/portfft/committed_descriptor_impl.hpp [hunk 24]
```diff
@@ -1058,20 +990,12 @@ class committed_descriptor_impl {
...
-    if (output_stride_0 < output_distance) {  // for example: batch interleaved output
-      output_distance = params.lengths.back();
-    }
 
     PORTFFT_LOG_TRACE("Dispatching the kernel for the last dimension");
     sycl::event previous_event =
         dispatch_kernel_1d(in, out, in_imag, out_imag, dependencies, params.number_of_transforms * outer_size,
-                           input_stride_0, output_stride_0, input_distance, output_distance, input_offset,
-                           output_offset, dimensions.back(), compute_direction);
+                           input_stride_0, output_stride_0, input_distance / outer_size, output_distance / outer_size,
```
[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/committed_descriptor_impl.hpp]: I've tried to clarify that here. https://github.com/codeplaysoftware/portFFT/pull/136/commits/3f12710323eb6a6a52d207a85da97d6af35d0467
The distance is the final dimension because the multidim transforms are always packed.

File: src/portfft/dispatcher/workitem_dispatcher.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/dispatcher/workitem_dispatcher.hpp]: done in https://github.com/codeplaysoftware/portFFT/pull/136/commits/f8a8661520105d38262d149a3686e87054807ba7

File: test/unit_test/instantiate_fft_tests.hpp (hunk 23 above)
Comment (FMarno) [reply test/unit_test/instantiate_fft_tests.hpp:272]: We agreed separately that we won't support that right now because of the test setup not supporting it.
Currently I have these tests for overlap
https://github.com/codeplaysoftware/portFFT/blob/fd9d7ccc5d6ec37fe673210c00a58ec5a159b7ff/test/unit_test/instantiate_fft_tests.hpp#L313-L318
the first one has an overlap in the forward domain
```
C++ exception with description "Domain forward: batch 1 collides with first batch at index 1" thrown in the test body.
```
the second has an overlap in the backward domain
```
C++ exception with description "Domain backward: batch 2 collides with first batch at index 6" thrown in the test body.
```
I'll update the readme too

File: src/portfft/committed_descriptor_impl.hpp (hunk 24 above)
[the remark this answers is not in the document]
Comment (Rbiessy) [reply src/portfft/committed_descriptor_impl.hpp]: I don't get how this works, the distance should be set to the total FFT size like we do here https://github.com/codeplaysoftware/portFFT/pull/136/files#diff-809aaf559d4b027a8bb3b0af31c4c9253a65cd4704510475f46acff32bc31478R142. This represents the number elements to move to get to the next batch. Does it have a different meaning for multi-dim FFTs that I forgot?

File: test/unit_test/instantiate_fft_tests.hpp (hunk 23 above)
Comment (Rbiessy) [reply test/unit_test/instantiate_fft_tests.hpp:272]: Leaving this open to clarify in the readme that overlap reads are not supported

File: src/portfft/committed_descriptor_impl.hpp (hunk 24 above)
[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/committed_descriptor_impl.hpp]: In this case we are submitting only the contiguous dimension of the multi-dim fft, and since the only layout is packed, we have a `params.number_of_transforms * outer_size` transforms of length `params.lengths.back()`.
If you look into `dispatch_kernel_1d` and then `dispatch_kernel_1d_helper` you'll see that the distance is only being used to decide which layout to dispatch to, so this can be removed when all the strided work is done.

[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/committed_descriptor_impl.hpp]: `in/output_stride/distance_0` are just being used to decide the layout to dispatch to here, so I've made that clearer by directly passing the layout in: https://github.com/codeplaysoftware/portFFT/pull/136/commits/3672ad563da0c6940ab52fd4c0d248bedf403e78

File: src/portfft/descriptor.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (Rbiessy) [reply src/portfft/descriptor.hpp:32]: I'm fine with either. Just a note that `descriptor_validate.hpp` will keep it close to `descriptor.hpp` when the files are listed in the IDE or GitHub so it's convenient too.

File: src/portfft/descriptor_validate.hpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/descriptor_validate.hpp:216]: I change my mind, you're correct, addressed here: https://github.com/codeplaysoftware/portFFT/pull/136/commits/c30f03068b9bdbfc8217642d2cbb1530bdaac25f

File: src/portfft/descriptor.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (FMarno) [reply src/portfft/descriptor.hpp]: I've changed it to descriptor_validation so hopefully get both qualities

**Review by Rbiessy — APPROVED** [review]

**Review by AD2605 — APPROVED** [review]
